File: hub/customer_voices.py
"""Staff-managed customer voice library shared by every ElevenLabs creator.

Only metadata is persisted. Cloning reservations survive timeouts so resubmitting
the same upload cannot create a second paid voice at the provider.
"""
from __future__ import annotations

import hashlib
import json
import os
import re
import uuid
from datetime import datetime, timezone

from flask import Blueprint, jsonify, render_template, request
from werkzeug.exceptions import RequestEntityTooLarge

from hub import jsonstore, audit
from hub.blueprint_guard import install

bp = Blueprint("customer_voices", __name__)
install(bp)
log = audit.for_module("customer_voices")
MAX_FILE = 25 * 1024 * 1024
MAX_TOTAL = 50 * 1024 * 1024
EXTENSIONS = {"mp3", "wav", "m4a", "aac", "ogg", "webm"}


class LibraryError(ValueError):
    pass


def _path():
    return os.path.join(jsonstore.data_root(), "customer-voices", "library.json")


def records():
    return list(jsonstore.read_json(_path(), default={}).values())


def public(row):
    return {k: row.get(k) for k in ("id", "voice_id", "name", "client", "description",
            "status", "preview_url", "created_at", "error")}


def ensure_usable(voice_id):
    """Stock voices pass through; managed voices must have cleared verification."""
    for row in records():
        if row.get("voice_id") == voice_id and row.get("status") != "ready":
            raise LibraryError("This customer voice is not ready. Open Customer Voices to check its status.")


def _save(record_id, changes):
    def change(rows):
        rows[record_id].update(changes)
        return rows
    jsonstore.update_json(_path(), change, default={})


def _status(voice):
    return "ready" if voice.get("requires_verification") is False else "verification_required"


@bp.get("/tools/customer-voices/")
def page():
    return render_template("customer_voices.html")


@bp.get("/api/customer-voices")
def listing():
    rows = records()
    if request.args.get("ready") == "1":
        rows = [r for r in rows if r.get("status") == "ready"]
    return jsonify(ok=True, voices=[public(r) for r in sorted(rows, key=lambda r: r.get("created_at", ""), reverse=True)])


def _capture_rows():
    from modules.commercial_builder.voice_capture_models import VoiceCaptureRequest
    query = VoiceCaptureRequest.query.filter_by(status="submitted", consent=True, revoked=False)
    capture_id = request.args.get("capture_id", type=int)
    if capture_id:
        query = query.filter_by(id=capture_id)
    return query.order_by(VoiceCaptureRequest.id.desc()).limit(100).all()


@bp.get("/api/customer-voices/captures")
def captures():
    return jsonify(ok=True, captures=[dict(id=r.id, client=r.client_name or (r.client.name if hasattr(r, "client") and r.client else ""),
        name=r.submitter_name, filename=r.original_filename, audio_url=r.audio_url) for r in _capture_rows()])


def _capture_sample(capture_id):
    from urllib.parse import urlsplit
    import requests
    from modules.commercial_builder.voice_capture_models import VoiceCaptureRequest
    row = VoiceCaptureRequest.query.get(capture_id)
    if not row or row.status != "submitted" or not row.consent or row.revoked:
        raise LibraryError("Choose a submitted recording with the speaker's permission.")
    url = urlsplit(row.audio_url or "")
    if url.scheme != "https" or url.hostname != "res.cloudinary.com" or url.username or url.password or url.port not in (None, 443):
        raise LibraryError("This recording cannot be imported directly. Download it from the capture panel and upload it here.")
    try:
        with requests.get(row.audio_url, timeout=(10, 45), stream=True, allow_redirects=False) as response:
            if response.status_code != 200:
                raise LibraryError("Could not read the submitted recording. Try again later.")
            data = bytearray()
            for chunk in response.iter_content(64 * 1024):
                data.extend(chunk)
                if len(data) > MAX_FILE:
                    raise LibraryError("Trim this recording to under 25 MB before uploading it here.")
    except requests.RequestException as exc:
        raise LibraryError("Could not read the submitted recording. Try again later.") from exc
    if not data:
        raise LibraryError("The submitted recording is empty.")
    return row, (row.original_filename or "recording.webm", bytes(data), row.mime_type or "audio/webm")


@bp.errorhandler(LibraryError)
def invalid(exc):
    return jsonify(ok=False, error=str(exc)), 400


@bp.errorhandler(RequestEntityTooLarge)
def too_large(exc):
    return jsonify(ok=False, error="Keep recordings under 25 MB each and 50 MB total."), 413


@bp.post("/api/customer-voices")
def create():
    from modules.radio_promo import voices
    request.max_content_length = MAX_TOTAL + 1024 * 1024
    form = request.form
    name, client = form.get("name", "").strip(), form.get("client", "").strip()
    description = form.get("description", "").strip()
    if not name or not client or len(name) > 120 or len(client) > 200 or len(description) > 1000:
        raise LibraryError("Enter a customer and voice name (up to 200 and 120 characters). Keep the description under 1,000 characters.")
    if form.get("authorized") != "true":
        raise LibraryError("Confirm that you own this voice or have the speaker's written permission.")
    record_id = form.get("request_id", "")
    try:
        record_id = str(uuid.UUID(record_id))
    except ValueError:
        raise LibraryError("Reload the page before submitting this voice.") from None
    existing_id = form.get("voice_id", "").strip()
    if existing_id and not re.fullmatch(r"[A-Za-z0-9_-]{1,100}", existing_id):
        raise LibraryError("Enter a valid ElevenLabs voice ID.")
    capture_id = form.get("capture_id", "").strip()
    samples = []
    uploads = request.files.getlist("files")
    if (existing_id and capture_id) or ((existing_id or capture_id) and uploads):
        raise LibraryError("Choose recordings or an existing voice ID.")
    if capture_id:
        if not capture_id.isdigit():
            raise LibraryError("Choose a submitted customer recording.")
        capture, sample = _capture_sample(int(capture_id))
        client = capture.client_name or client
        samples.append(sample)
    if not existing_id and not capture_id and not 1 <= len(uploads) <= 5:
        raise LibraryError("Upload between one and five recordings.")
    total = 0
    for upload in uploads:
        filename = (upload.filename or "").replace("\\", "/").rsplit("/", 1)[-1]
        if filename.rsplit(".", 1)[-1].lower() not in EXTENSIONS:
            raise LibraryError("Use MP3, WAV, M4A, AAC, OGG, or WebM recordings.")
        data = upload.read(MAX_FILE + 1)
        total += len(data)
        if not data or len(data) > MAX_FILE or total > MAX_TOTAL:
            raise LibraryError("Recordings must be nonempty, under 25 MB each and 50 MB total.")
        samples.append((filename, data, upload.mimetype))
    fingerprint = hashlib.sha256(json.dumps([name, client, description, existing_id,
        [hashlib.sha256(s[1]).hexdigest() for s in samples]], sort_keys=True).encode()).hexdigest()
    from hub.auth import user_from_environ
    user = user_from_environ(request.environ) or {}
    row = dict(id=record_id, name=name, client=client, description=description, voice_id=existing_id or None,
               status="creating", created_at=datetime.now(timezone.utc).isoformat(),
               fingerprint=fingerprint, capture_id=capture_id or None, permission_confirmed=True,
               permission_by=str(user or "staff"))
    prior = []

    def reserve(rows):
        if record_id in rows:
            old = rows[record_id]
            if old.get("status") != "failed" or old.get("fingerprint") != fingerprint:
                prior.append(old)
                return None
            # A retry still checks other tabs for an accepted or in-flight clone.
        if not existing_id:
            matching = next((r for r in rows.values() if r.get("fingerprint") == fingerprint and r.get("status") != "failed"), None)
            if matching:
                prior.append(matching)
                return None
        if existing_id and any(r.get("voice_id") == existing_id and r.get("status") not in ("failed", "needs_review") for r in rows.values()):
            raise LibraryError("That voice is already saved in Customer Voices.")
        if existing_id:
            for key in list(rows):
                if rows[key].get("voice_id") == existing_id and rows[key].get("status") in ("failed", "needs_review"):
                    del rows[key]  # Replace a failed read-only import, preserving one guard per voice.
        rows[record_id] = row
        return rows

    jsonstore.update_json(_path(), reserve, default={})
    if prior:
        old = prior[0]
        if old.get("fingerprint") != fingerprint:
            return jsonify(ok=False, error="This submission already exists. Start a new voice for different recordings."), 409
        if old.get("status") in ("creating", "needs_review"):
            return jsonify(ok=False, voice=public(old), error="This submission is awaiting confirmation. Check ElevenLabs before starting another clone; save its existing voice ID if it completed."), 409
        return jsonify(ok=True, voice=public(old))
    try:
        voice = voices.get_voice(existing_id) if existing_id else voices.clone_voice(name, samples, description)
    except (voices.VoiceError, ValueError) as exc:
        safe_to_retry = bool(existing_id) or isinstance(exc, voices.VoiceRequestRejected)
        message = str(exc) if isinstance(exc, voices.VoiceError) else "ElevenLabs returned an unreadable voice response."
        if not safe_to_retry:
            message += " The result is uncertain. Check ElevenLabs and use its existing voice ID if it was created; this submission will not create another clone."
        _save(record_id, dict(status="failed" if safe_to_retry else "needs_review", error=message))
        return jsonify(ok=False, error=message), 502
    changes = dict(voice_id=voice["voice_id"], status=_status(voice), preview_url=voice.get("preview_url") or "", error=None)
    _save(record_id, changes)
    row.update(changes)
    log("voice_saved", client=client, detail=name, voice_id=row["voice_id"])
    return jsonify(ok=True, voice=public(row)), 201


@bp.post("/api/customer-voices/<record_id>/refresh")
def refresh(record_id):
    from modules.radio_promo import voices
    row = next((r for r in records() if r["id"] == record_id), None)
    if not row or not row.get("voice_id"):
        raise LibraryError("This submission has no confirmed voice ID. Check ElevenLabs and save the existing voice ID if it completed.")
    try:
        voice = voices.get_voice(row["voice_id"])
    except voices.VoiceError as exc:
        return jsonify(ok=False, error=str(exc)), 502
    changes = dict(status=_status(voice), preview_url=voice.get("preview_url") or "", error=None)
    _save(record_id, changes)
    row.update(changes)
    log("voice_refreshed", client=row["client"], detail=row["name"], voice_id=row["voice_id"])
    return jsonify(ok=True, voice=public(row))

File: hub/static/customer-voices.js
/* Staff voice library. Never retain the customer's sample recordings in browser storage. */
(() => {
  const form = document.getElementById('cv-form');
  const get = id => document.getElementById(id);
  const states = {ready:'Ready to use', verification_required:'Verification required', creating:'Awaiting confirmation', needs_review:'Confirmation needed', failed:'Not saved — retry after correction'};
  let requestId = crypto.randomUUID(), recoveryConfirmation = null;
  async function api(url, options) {
    const res = await fetch(url, options); let data;
    try { data = await res.json(); } catch (_) { throw new Error('The server did not confirm this request. Refresh the library before trying again.'); }
    if (!res.ok || !data.ok) throw new Error(data.error || 'Could not complete this request.');
    return data;
  }
  async function load() {
    const list = get('cv-list'); list.textContent = 'Loading voices…';
    try {
      const {voices} = await api('/api/customer-voices'); list.replaceChildren();
      if (!voices.length) list.textContent = 'No customer voices yet. Add your first voice here.';
      voices.forEach(v => {
        const row = document.createElement('article'); row.className = 'cv-voice';
        const title = document.createElement('h3'); title.textContent = `${v.name} · ${v.client}`;
        const status = document.createElement('p'); status.className = 'cv-status'; status.textContent = states[v.status] || v.status;
        row.append(title, status);
        if (v.error) { const p = document.createElement('p'); p.textContent = v.error; row.append(p); }
        if (v.preview_url && /^https:\/\//i.test(v.preview_url)) { const audio = document.createElement('audio'); audio.controls = true; audio.preload = 'none'; audio.src = v.preview_url; row.append(audio); }
        if (v.voice_id) {
          const button = document.createElement('button'); button.type = 'button'; button.textContent = 'Refresh voice status';
          button.onclick = async () => { button.disabled = true; try { await api(`/api/customer-voices/${encodeURIComponent(v.id)}/refresh`, {method:'POST'}); await load(); } catch(e) { status.textContent = e.message; button.disabled = false; } };
          row.append(button);
        }
        if (v.status === 'needs_review' && !v.voice_id) {
          const retry = document.createElement('button'); retry.type = 'button'; retry.textContent = 'Retry upload after checking ElevenLabs';
          retry.onclick = () => {
            if (!window.confirm('Have you checked the connected ElevenLabs account and confirmed this voice was NOT created? Continuing allows a new cloning attempt.')) return;
            requestId = v.id; recoveryConfirmation = crypto.randomUUID(); form.reset(); get('cv-mode').value = 'upload'; get('cv-mode').onchange();
            get('cv-name').value = v.name || ''; get('cv-client').value = v.client || ''; get('cv-description').value = v.description || '';
            get('cv-submit').disabled = false; get('cv-message').textContent = 'Select the same recordings, confirm permission, then create the voice again.'; get('cv-files').focus();
          };
          row.append(retry);
        }
        if (v.status === 'needs_review' || v.status === 'failed') {
          const recover = document.createElement('button'); recover.type = 'button'; recover.textContent = 'Use existing ElevenLabs voice ID';
          recover.onclick = () => {
            requestId = crypto.randomUUID(); recoveryConfirmation = null; form.reset(); get('cv-mode').value = 'existing'; get('cv-mode').onchange();
            get('cv-name').value = v.name || ''; get('cv-client').value = v.client || ''; get('cv-id').value = v.voice_id || '';
            get('cv-submit').disabled = false; get('cv-message').textContent = 'Paste the voice ID from ElevenLabs to save the existing voice without creating another clone.';
            get('cv-id').focus();
          };
          row.append(recover);
        }
        list.append(row);
      });
    } catch(e) { list.textContent = e.message; }
  }
  get('cv-client').value = new URLSearchParams(location.search).get('client') || '';
  let captureRows = [];
  get('cv-mode').onchange = async () => {
    const mode = get('cv-mode').value, existing = mode === 'existing', capture = mode === 'capture';
    get('cv-upload').hidden = existing || capture; get('cv-existing').hidden = !existing; get('cv-capture').hidden = !capture;
    get('cv-files').disabled = existing || capture; get('cv-files').required = !existing && !capture;
    get('cv-id').disabled = !existing; get('cv-id').required = existing;
    get('cv-capture-id').disabled = !capture; get('cv-capture-id').required = capture;
    get('cv-submit').textContent = existing ? 'Save customer voice' : 'Create customer voice';
    if (capture) {
      try {
        const requestedCapture = new URLSearchParams(location.search).get('capture_id');
        const data = await api('/api/customer-voices/captures' + (requestedCapture ? '?capture_id=' + encodeURIComponent(requestedCapture) : '')); captureRows = data.captures || [];
        const select = get('cv-capture-id'); select.replaceChildren(new Option('Choose a submitted recording', ''));
        captureRows.forEach(r => select.add(new Option(`${r.client} — ${r.name || r.filename}`, r.id)));
        get('cv-capture-note').textContent = captureRows.length ? 'Listen to the submitted recording before creating its reusable voice.' : 'No submitted recordings yet. Create a client recording link in Commercial Builder, or upload recordings here.';
        const requested = new URLSearchParams(location.search).get('capture_id');
        if (requested) { select.value = requested; select.onchange(); }
      } catch(e) { get('cv-capture-note').textContent = e.message; }
    }
  };
  get('cv-capture-id').onchange = () => {
    const row = captureRows.find(r => String(r.id) === get('cv-capture-id').value);
    const audio = get('cv-capture-audio'); audio.pause(); audio.hidden = !row?.audio_url;
    if (row) { get('cv-client').value = row.client || ''; get('cv-name').value = row.name || ''; if (row.audio_url) audio.src = row.audio_url; }
  };
  if (new URLSearchParams(location.search).has('capture_id')) { get('cv-mode').value = 'capture'; get('cv-mode').onchange(); }
  form.onsubmit = async e => {
    e.preventDefault(); const button = get('cv-submit'), message = get('cv-message');
    const files = get('cv-files').disabled ? [] : [...get('cv-files').files];
    if (files.length > 5 || files.some(f => !f.size || f.size > 25*1024*1024) || files.reduce((n,f) => n+f.size,0) > 50*1024*1024) { message.textContent = 'Use 1–5 nonempty recordings, at most 25 MB each and 50 MB total.'; return; }
    button.disabled = true; message.textContent = 'Saving voice… This can take a few minutes.';
    const body = new FormData(form); body.set('authorized', 'true'); body.set('request_id', requestId);
    if (recoveryConfirmation) body.set('recovery_confirmation', recoveryConfirmation);
    recoveryConfirmation = null;
    try {
      const {voice} = await api('/api/customer-voices', {method:'POST', body});
      message.textContent = voice.status === 'ready' ? 'Saved. This voice is now available in all speech creators.' : 'Saved. Complete verification in ElevenLabs, then refresh its status.';
      get('cv-files').value = ''; get('cv-new').hidden = false;
    } catch(e) { message.textContent = e.message; button.disabled = false; get('cv-new').hidden = false; }
    await load();
  };
  get('cv-new').onclick = () => { requestId = crypto.randomUUID(); recoveryConfirmation = null; form.reset(); get('cv-mode').onchange(); get('cv-submit').disabled = false; get('cv-new').hidden = true; get('cv-message').textContent = ''; };
  get('cv-reload').onclick = load;
  load();
})();
